src/libraries/System.Private.CoreLib/src/System/WeakReference.T.cs

```diff
@@ -13,7 +15,19 @@ namespace System
     public sealed partial class WeakReference<T> : ISerializable
         where T : class?
     {
-        // If you fix bugs here, please fix them in WeakReference at the same time.
+        // If you fix bugs here, please fix them in WeakReference<T> at the same time.
+
+        // Most methods using the handle should use GC.KeepAlive(this) to avoid potential handle recycling
+        // attacks (i.e. if the WeakReference instance is finalized away underneath you when you're still
+        // handling a cached value of the handle then the handle could be freed and reused).
+
+        // the handle field is effectively readonly until the object is finalized.
+        // CoreCLR specific: The implementation is in native code and the field is not a regular GC handle.
+        //                   It can have a special values that are used to prevent a race condition between setting the target and finalization.
+        private IntPtr _handleAndKind;
+
+        // the lowermost bit is used to indicate whether the handle is tracking resurrection
+        private const nint TracksResurrectionBit = 1;
```

Comment:
```suggestion
#if !CORECLR
        // the handle field is effectively readonly until the object is finalized.
        private IntPtr _handleAndKind;

        // the lowermost bit is used to indicate whether the handle is tracking resurrection
        private const nint TracksResurrectionBit = 1;
#endif
```
This will fix the build break, and also reduces confusion.

> It can have a special values that are used to prevent a race condition between setting the target and finalization.

The primary reason for having the special value and spin lock are race conditions between setting the target from two different threads in the presence of the special COM/WinRT support.

Reply:
I tried to use just one handle field in all implementations, but it looks like CorCLR has dependency on `m_handle`.

Reply:
I still need to figure the ways of COM/WinRT support in CorCLR. 

It looks like NativeAOT had support for that too, which was simpler for the path that does not engage with COM, but the code is gone. Maybe we have in some other repo.

Reply:
I do not think that the code was ever open sourced. You can find it in the ProjectN branch. It does not have good performance characteristics so I do not think it would be usable as is. Here is the problem:

The COM support needs to attach an extra bit of information to the WeakReference instance, and take slower paths to get the target when this extra bit of information is attached to the WeakReference instance. The straightforward ways to do that would be an extra field, or attach it via ConditionalWeakTable, or attach it via syncblock (it is what CoreCLR does in native code - it is the internal variant of ConditionalWeakTable). The problem with extra field is an extra footprint. The problem with ConditionalWeakTable or syncblock is extra path length. We would have to do ConditionalWeakTable/SyncBlock lookup everytime even if there is no COM involved at all. The CoreCLR solution is to encode whether the given WeakReference instance  needs to go through the COM path in the low bit of the handle and use it as a quick check that says whether to take the slow COM path; or whether the simple fast path is ok.

Does it make sense?

Reply:
> Does it make sense?

Kind of. It looks like the whole deal is keyed off IWeakReference, otherwise RCW is the same as a regular .NET object. I assume IWeakReference is rare, so ideally supporting it would be pay-for-play. In such case some scheme combining ConditionalWeakTable and a bit on the handle would be a promising approach.

I really do not like the part where CoreCLR uses a special kind of a handle to maybe store some extra data. As a result Set may need to change the kind of the handle, thus Set is not atomic and needs to take a spinlock and then getter needs to be aware of that lock and finalizer is complicated too - everyone pays a steep price for a possibility that one day we see a IWeakReference to come by.

I will examine ProjectN approach. Even if not the most performant for COM/IWeakReference case, it may be better for non-COM case, which I think we should optimize for.

Reply:
>I will examine ProjectN approach. Even if not the most performant for COM/IWeakReference case, it may be better for non-COM case, which I think we should optimize for.

ProjectN approach is not performant for the non-COM case.